### ISSUE TITLE
Rule Request: Static Operator

### ISSUE THREAD

**Issue by marcelofabri** [realm/SwiftLint#2395]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### New rule request

Operators should be declared as static functions, not free functions. Operators used to be declared as free functions but at some point (Swift 3?) they can also be declared as `static` methods on the type that they operate.

1. Why should this rule be added? Share links to existing discussion about what
   the community thinks about this.

https://docs.swift.org/swift-book/LanguageGuide/AdvancedOperators.html#ID42

2. Provide several examples of what _would_ and _wouldn't_ trigger violations.

<details>
<summary>Non Triggering Examples</summary>

```swift
class A: Equatable {
    static func == (lhs: A, rhs: A) -> Bool {
        return false
    }
```

```swift
class A<T>: Equatable {
    static func == <T>(lhs: A<T>, rhs: A<T>) -> Bool {
        return false
    }
```

```swift
public extension Array where Element == Rule {
    static func == (lhs: Array, rhs: Array) -> Bool {
        if lhs.count != rhs.count { return false }
        return !zip(lhs, rhs).contains { !$0.0.isEqualTo($0.1) }
    }
}
```

```swift
private extension Optional where Wrapped: Comparable {
    static func < (lhs: Optional, rhs: Optional) -> Bool {
        switch (lhs, rhs) {
        case let (lhs?, rhs?):
            return lhs < rhs
        case (nil, _?):
            return true
        default:
            return false
        }
    }
}
```

</details>
<details>
<summary>Triggering Examples</summary>

```swift
↓func == (lhs: A, rhs: A) -> Bool {
    return false
}
```

```swift
↓func == <T>(lhs: A<T>, rhs: A<T>) -> Bool {
    return false
}
```

```swift
↓func == (lhs: [Rule], rhs: [Rule]) -> Bool {
    if lhs.count != rhs.count { return false }
    return !zip(lhs, rhs).contains { !$0.0.isEqualTo($0.1) }
}
```

```swift
private ↓func < <T: Comparable>(lhs: T?, rhs: T?) -> Bool {
    switch (lhs, rhs) {
    case let (lhs?, rhs?):
        return lhs < rhs
    case (nil, _?):
        return true
    default:
        return false
    }
}
```

</details>

3. Should the rule be configurable, if so what parameters should be configurable?

Only severity.

4. Should the rule be opt-in or enabled by default? Why?
   See [README.md](../README.md#opt-in-rules) for guidelines on when to mark a rule as opt-in.

Opt-in as a lot of code still uses free functions. Potentially there're also cases where declaring in a type is not possible.
